keps/sig-apps/4443-configurable-pod-failure-policy-reasons/README.md

```diff
@@ -0,0 +1,927 @@
+  - name: simulations
+    replicas: 10
+    template:
+      spec:
+        parallelism: 1
+        completions: 1
+        backoffLimit: 0
+        # If a pod fails with exit code 2 or 3, fail the Job, using the user-defined reason.
+        podFailurePolicy:
+          rules:
+          - action: FailJob
+            onExitCodes:
+              containerName: main
+              operator: In
+              values: [2]
+            reason: "ExitCode2"
+          - action: FailJob
+            onExitCodes:
+              containerName: main
+              operator: In
+              values: [3]
+            reason: "ExitCode3"
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: python:3.10
+              command: ["..."]
+```
+
+**Alternative example: single Job with a Pod Failure Policy configuration for this use case**:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: simulation
+spec:
+  parallelism: 1
+  completions: 1
+  backoffLimit: 0
+  # If a pod fails with exit code 2 or 3, fail the Job, using the user-defined reason.
+  podFailurePolicy:
+    rules:
+    - action: FailJob
+      onExitCodes:
+        containerName: main
+        operator: In
+        values: [2]
+      reason: "ExitCode2"
+    - action: FailJob
+      onExitCodes:
+        containerName: main
+        operator: In
+        values: [3]
+      reason: "ExitCode3"
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: main
+        image: python:3.10
+        command: ["..."]
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+I don't see any notable risks for this feature.
+
+
+## Design Details
+
+If unset, it will default to `PodFailurePolicy`, which is the current [reason](https://sourcegraph.com/github.com/kubernetes/kubernetes@6a4e93e776a35d14a61244185c848c3b5832621c/-/blob/staging/src/k8s.io/api/batch/v1/types.go?L542) the Job
+controller uses when a PodFailurePolicy triggers a Job failure.
+
+When a `PodFailurePolicyRule` matches a pod failure and the `Action` is `FailJob`, the Job
+controller will add the reason defined in the `Reason` field to the JobFailed [condition](https://sourcegraph.com/github.com/kubernetes/kubernetes@6a4e93e776a35d14a61244185c848c3b5832621c/-/blob/pkg/controller/job/job_controller.go?L816) added
```

Comment:
No strong view, but I think it could be preferable to let the users only specify the suffix for the reason. This would still support the use cases, but having the PodFailurePolicy prefix might be handy when debugging the job by another person than the user.

Regardless, I would like to align the approach with https://github.com/kubernetes/enhancements/pull/4062, because this is pretty much the same use case.

Reply:
-1 to a suffix

This would complicate the usage of the feature. In the example in the story, the user would have to put different strings in the spec of the jobset and the job.

Reply:
I documented "the entire reason" vs "the suffix of the reason" here: https://github.com/kubernetes/enhancements/blob/bfd01d4f3668c3dc883e59b8f4feb06ff8de100f/keps/sig-apps/3998-job-success-completion-policy/README.md#possibility-for-the-configurable-reason-for-the-successcriteriamet-condition

This may be helpful.

Reply:
In #4062 it sounds like we aren't going to add a Reason field to the Success Policy, so we don't need to align the approach here anymore.

For our case, I do think "the entire reason" is preferable to "suffix of the reason" for the reason @alculquicondor mentioned: I can see the argument for having a consistent prefix, but I think a k8s-defined prefix being prepended to the user-defined reason in an opaque manner would cause bugs and result in a more confusing user experience. 

As a user, I would intuitively expect to be able to "link" the JobSet FailurePolicyRule "Reason" and the child Job's PodFailurePolicyRule "Reason"  by setting them to the same value. To me, it would be non-obvious and unexpected if the Reason on the `JobFailed` condition was not exactly what I had defined in my PodFailurePolicyRule.

Reply:
I don't like the suffix as well, it is easier to the user to simply use the exact same string in two places

Reply:
I agree a suffix is not ideal but this was the final design we managed to achieve consensus on in the WG batch meeting. Going to resolve this thread now but feel free to reopen if you think there is more to discuss.